### PR TITLE
Feature/Refactor getDeployTransaction

### PR DESCRIPTION
**Pull request by natanasow**
**Description**:
The method should be refactored such that it returns an array of TransactionRequests:
- first element returned should be the FileCreate TX
- n number of File Appends
- last element should be the ContractCreate TX

**Related issue(s)**:

Fixes #
- https://github.com/hashgraph/hethers.js/issues/81
**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by joan41868 — COMMENTED** [review]
Looks great. I've left some comments :)

File: packages/contracts/src.ts/index.ts [hunk 1]
```diff
@@ -1177,46 +1178,34 @@ export class ContractFactory {
...
-        logger.checkArgumentCount(args.length, this.interface.deploy.inputs.length, " in Contract constructor");
-
-        // Set the data to the bytecode + the encoded constructor arguments
-        tx.data = hexlify(concat([
-            this.bytecode,
-            this.interface.encodeDeploy(args)
-        ]));
+        const contractCreate: TransactionRequest = {
+            ...args,
+            customData: {}
```
Comment (joan41868) [inline packages/contracts/src.ts/index.ts:1205]: We should add `gasLimit` property here. I suggest we hardcode 300k for now

**Comment by joan41868** [comment]
You shoud also pull the latest `develop` state in your branch.

**Review by joan41868 — APPROVED** [review]

File: packages/contracts/src.ts/index.ts [hunk 2]
```diff
@@ -1177,46 +1178,33 @@ export class ContractFactory {
...
+    getDeployTransactions(...args: Array<any>): Array<TransactionRequest> {
+        let chunks = splitInChunks(Buffer.from(this.bytecode).toString(), 4096);
 
-        // If we have 1 additional argument, we allow transaction overrides
-        if (args.length === this.interface.deploy.inputs.length + 1 && typeof(args[args.length - 1]) === "object") {
-            tx = shallowCopy(args.pop());
-            for (const key in tx) {
-                if (!allowedTransactionKeys[key]) {
-                    throw new Error("unknown transaction override " + key);
-                }
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:1190]: We should leave this check. The reason why it exists is that the user can provide an additional argument `transaction overrides` as part of **every** transaction. 

The overrides that can be passed are:
- `value`
- `gasLimit`

image

File: packages/contracts/src.ts/index.ts [hunk 3]
```diff
@@ -1177,46 +1178,33 @@ export class ContractFactory {
...
+                fileChunk: chunks[0]
             }
-        }
+        };
 
-        // Do not allow these to be overridden in a deployment transaction
-        ["data", "from", "to"].forEach((key) => {
-            if ((<any>tx)[key] == null) { return; }
-            logger.throwError("cannot override " + key, Logger.errors.UNSUPPORTED_OPERATION, { operation: key })
-        });
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:1198]: We should leave this check as-well

File: packages/contracts/src.ts/index.ts [hunk 4]
```diff
@@ -1177,46 +1178,33 @@ export class ContractFactory {
...
+            };
 
-        if (tx.value) {
-            const value = BigNumber.from(tx.value);
-            if (!value.isZero() && !this.interface.deploy.payable) {
-                logger.throwError("non-payable constructor cannot override value", Logger.errors.UNSUPPORTED_OPERATION, {
-                    operation: "overrides.value",
-                    value: tx.value
-                });
-            }
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:1207]: This check must be kept as-well. This is a safe guard for users. If they transfer value on contract deployment which does not have `payable` set in the constructor the deployment will fail

File: packages/contracts/src.ts/index.ts [hunk 5]
```diff
@@ -1177,46 +1178,33 @@ export class ContractFactory {
...
+            fileAppends.push(fileAppend);
         }
 
-        // Make sure the call matches the constructor signature
-        logger.checkArgumentCount(args.length, this.interface.deploy.inputs.length, " in Contract constructor");
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:1211]: This check should be kept as-well

File: packages/contracts/src.ts/index.ts [hunk 6]
```diff
@@ -1177,46 +1178,33 @@ export class ContractFactory {
...
-
-        // Set the data to the bytecode + the encoded constructor arguments
-        tx.data = hexlify(concat([
-            this.bytecode,
-            this.interface.encodeDeploy(args)
-        ]));
+        const contractCreate: TransactionRequest = {
+            gasLimit: 300000,
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:1202]: The gas limit must not be hardcoded here. It must be passed as `transaction override` from the outside

File: packages/contracts/src.ts/index.ts [hunk 7]
```diff
@@ -1236,10 +1224,10 @@ export class ContractFactory {
         params.push(overrides);
 
         // Get the deployment transaction (with optional overrides)
-        const unsignedTx = this.getDeployTransaction(...params);
+        const unsignedTx = this.getDeployTransactions();
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:1227]: We should pass the `...params` otherwise we won't be able to deploy a contract with constructor arguments or pass the `gasLimit`

File: packages/tests/src.ts/test-contract.spec.ts [hunk 8]
```diff
@@ -291,6 +291,21 @@ describe("Test Contract Transaction Population", function() {
...
+        };
+        const provider = ethers.providers.getDefaultProvider('previewnet');
+        // @ts-ignore
+        const wallet = new ethers.Wallet(hederaEoa, provider);
+
+        const contractFactory = new ethers.ContractFactory(abi, "", wallet);
+        const transactions = contractFactory.getDeployTransactions();
+
+        assert.strictEqual(Array.isArray(transactions), true);
+        assert.strictEqual(transactions.length, 2);
```
Comment (Daniel-K-Ivanov) [inline packages/tests/src.ts/test-contract.spec.ts:307]: We should assert the exact properties of the transactions not only that they are an array of 2 elements

File: packages/contracts/src.ts/index.ts [hunk 9]
```diff
@@ -106,12 +107,12 @@ export interface ContractTransaction extends TransactionResponse {
...
-    type: true, accessList: true,
-    maxFeePerGas: true, maxPriorityFeePerGas: true,
-    customData: true
-}
+// const allowedTransactionKeys: { [ key: string ]: boolean } = {
+//     chainId: true, data: true, from: true, gasLimit: true, gasPrice:true, nonce: true, to: true, value: true,
+//     type: true, accessList: true,
+//     maxFeePerGas: true, maxPriorityFeePerGas: true,
+//     customData: true
+// }
```
Comment (Daniel-K-Ivanov) [inline packages/contracts/src.ts/index.ts:115]: Should be uncommented

File: packages/abstract-signer/src.ts/index.ts [hunk 10]
```diff
@@ -23,6 +23,7 @@ import {
 } from "@hashgraph/sdk";
 import * as Long from "long";
 import { SignedTransaction, TransactionBody } from "@hashgraph/proto";
+import { splitInChunks } from "@ethersproject/strings";
```
Comment (Daniel-K-Ivanov) [inline packages/abstract-signer/src.ts/index.ts:26]: I think that `splitInChunks` is better to be in the `bytes` package and to use only `bytes` instead of `string`.

File: packages/transactions/src.ts/index.ts [hunk 11]
```diff
@@ -356,7 +356,7 @@ export function serialize(transaction: UnsignedTransaction, signature?: Signatur
     });
 }
 
-export function serializeHederaTransaction(transaction: TransactionRequest) : HederaTransaction {
+export function serializeHederaTransaction(transaction: TransactionRequest, pubKey?: HederaPubKey) : HederaTransaction {
```
Comment (Daniel-K-Ivanov) [inline packages/transactions/src.ts/index.ts:359]: Why do we need to add `pubkey` as another argument in the `serialiseHederaTransaction`?

File: packages/transactions/src.ts/index.ts [hunk 12]
```diff
@@ -384,9 +384,10 @@ export function serializeHederaTransaction(transaction: TransactionRequest) : He
                 // only a chunk, thus the first one
                 tx = new FileCreateTransaction()
                     .setContents(transaction.customData.fileChunk)
-                    .setKeys([ transaction.customData.fileKey ?
+                    .setKeys([transaction.customData.fileKey ?
                         transaction.customData.fileKey :
-                        HederaPubKey.fromString(this._signingKey().compressedPublicKey) ])
+                        pubKey
```
Comment (Daniel-K-Ivanov) [inline packages/transactions/src.ts/index.ts:389]: Why is this change necessary?
